runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -275,13 +277,53 @@ private boolean matches(MetricsFilter filter, MetricKey key) {
         && matchesScope(key.stepName(), filter.steps());
   }
 
-  private boolean matchesScope(String actualScope, Set<String> scopes) {
+  /** Convert a list of Strings into a Map of String to its index in the list. */
+  private Map<String, List<Integer>> makeScopeIndex(String[] scopeSplit) {
+    Map<String, List<Integer>> result = new HashMap<String, List<Integer>>();
+    for (int i = 0; i < scopeSplit.length; i++) {
+      String stage = scopeSplit[i];
+      if (!result.containsKey(stage)) {
+        result.put(stage, new ArrayList<Integer>());
+      }
+      result.get(stage).add(i);
+    }
+    return result;
+  }
+
+  /** Realize a substring search between the actual scope and the filter scope. */
+  private boolean matchesDeepScope(String[] actualScopeSplit,
```

Comment:
This looks good, but let's try to clean up the description of "substring" search, and maybe make the implementation more straightforward. Consider:

```
/**
  * {@code deepMatch(needle, haystack)} returns true if {@code needle} represents a path within
  * {@code haystack}. For example, "foo/bar" is in "a/foo/bar/b", but not "a/fool/bar/b" or
  * "a/foo/bart/b".
  */
private boolean deepMatch(String needle, String haystack) {
  int location = haystack.indexOf(needle);
  int end = location + needle.length();
  if (location == -1) {
    return false;  // needle not found
  } else if (location != 0 && haystack.charAt(location - 1) != '/') {
    return false; // the first entry in needle wasn't exactly matched
  } else (end != haystack.length() && haystack.charAt(end) != '/') {
    return false; // the last entry in needle wasn't exactly matched
  }  else {
    return true;
  }
} 
```

I think that implementation should be the same. If not, let me know and I'll send some comments on how to cleanup the current approach.